### ISSUE TITLE
w_subset(weights:W, ids:np.ndarray) constructs faulty weights object

### ISSUE THREAD

**Issue by jtsmn** [pysal/pysal#799]
- Platform information:
```
>>> import os; print(os.name, os.sys.platform);print(os.uname())
>>> posix darwin
posix.uname_result(sysname='Darwin', nodename='innsbruck-umh.cs.umn.edu', release='15.3.0', version='Darwin Kernel Version 15.3.0: Thu Dec 10 18:40:58 PST 2015; root:xnu-3248.30.4~1/RELEASE_X86_64', machine='x86_64')
```
- Python version: 
```
>>> 3.4.4 (default, Mar  2 2016, 03:31:27) 
[GCC 4.2.1 Compatible Apple LLVM 7.0.2 (clang-700.1.81)]
```
- SciPy version:
```
>>> import scipy; print(scipy.__version__)
>>> 0.17.0
```
- NumPy version:
```
>>> import numpy; print(numpy.__version__)
>>> 1.10.4
```
When I try to run `w_subset()`, I see the following error:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-377-f4f0fc2561a1> in <module>()
     36 dvs.append(ps.lag_spatial(yellow_w, yellow['hinc_10k'].values))
     37 dvs = np.array(dvs).T
---> 38 lag = ps.spreg.ml_lag.ML_Lag(iv, dvs, yellow_w, name_y='# of Rides', name_x=['% White', 'W_% White', '% Owned', 'W_% Owned', 'Median HINC', 'W_Median HINC'])
     39 print(lag.summary)

/Users/theba004/Dropbox/Work/Development/venvs/py3/jupyter/lib/python3.4/site-packages/pysal/spreg/ml_lag.py in __init__(self, y, x, w, method, epsilon, spat_diag, vm, name_y, name_x, name_w, name_ds)
    542         method = method.upper()
    543         BaseML_Lag.__init__(
--> 544             self, y=y, x=x_constant, w=w, method=method, epsilon=epsilon)
    545         # increase by 1 to have correct aic and sc, include rho in count
    546         self.k += 1

/Users/theba004/Dropbox/Work/Development/venvs/py3/jupyter/lib/python3.4/site-packages/pysal/spreg/ml_lag.py in __init__(self, y, x, w, method, epsilon)
    195         if methodML in ['FULL', 'LU', 'ORD']:
    196             if methodML == 'FULL':
--> 197                 W = w.full()[0]     # moved here
    198                 res = minimize_scalar(lag_c_loglik, 0.0, bounds=(-1.0, 1.0),
    199                                       args=(

/Users/theba004/Dropbox/Work/Development/venvs/py3/jupyter/lib/python3.4/site-packages/pysal/weights/weights.py in full(self)
    952 
    953         """
--> 954         return util.full(self)
    955 
    956     def towsp(self):

/Users/theba004/Dropbox/Work/Development/venvs/py3/jupyter/lib/python3.4/site-packages/pysal/weights/util.py in full(w)
    704     wfull = np.zeros([w.n, w.n], dtype=float)
    705     keys = list(w.neighbors.keys())
--> 706     if w.id_order:
    707         keys = w.id_order
    708     for i, key in enumerate(keys):

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

Which to me suggests that the [check](https://github.com/pysal/pysal/blob/master/pysal/weights/util.py#L705) for having a value of `w.id_order` isn't working the way this if-statement is intended. A quick skim of `util.py` suggests you might want:

```
if w.id_order is not None:
   keys = w.id_order
```

**Comment by jtsmn** [comment]
In fact, 
```
if w.id_order.any():
        keys = w.id_order
```
does allow it to pass line 706.

However, the subsequent error:
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-10-f4f0fc2561a1> in <module>()
     36 dvs.append(ps.lag_spatial(yellow_w, yellow['hinc_10k'].values))
     37 dvs = np.array(dvs).T
---> 38 lag = ps.spreg.ml_lag.ML_Lag(iv, dvs, yellow_w, name_y='# of Rides', name_x=['% White', 'W_% White', '% Owned', 'W_% Owned', 'Median HINC', 'W_Median HINC'])
     39 print(lag.summary)

/Users/theba004/Dropbox/Work/Development/venvs/py3/jupyter/lib/python3.4/site-packages/pysal/spreg/ml_lag.py in __init__(self, y, x, w, method, epsilon, spat_diag, vm, name_y, name_x, name_w, name_ds)
    542         method = method.upper()
    543         BaseML_Lag.__init__(
--> 544             self, y=y, x=x_constant, w=w, method=method, epsilon=epsilon)
    545         # increase by 1 to have correct aic and sc, include rho in count
    546         self.k += 1

/Users/theba004/Dropbox/Work/Development/venvs/py3/jupyter/lib/python3.4/site-packages/pysal/spreg/ml_lag.py in __init__(self, y, x, w, method, epsilon)
    195         if methodML in ['FULL', 'LU', 'ORD']:
    196             if methodML == 'FULL':
--> 197                 W = w.full()[0]     # moved here
    198                 res = minimize_scalar(lag_c_loglik, 0.0, bounds=(-1.0, 1.0),
    199                                       args=(

/Users/theba004/Dropbox/Work/Development/venvs/py3/jupyter/lib/python3.4/site-packages/pysal/weights/weights.py in full(self)
    952 
    953         """
--> 954         return util.full(self)
    955 
    956     def towsp(self):

/Users/theba004/Dropbox/Work/Development/venvs/py3/jupyter/lib/python3.4/site-packages/pysal/weights/util.py in full(w)
    710         w_i = w.weights[key]
    711         for j, wij in zip(n_i, w_i):
--> 712             c = keys.index(j)
    713             wfull[i, c] = wij
    714     return (wfull, keys)

AttributeError: 'numpy.ndarray' object has no attribute 'index'
```
suggests there may be something else going on with `w.id_order`

**Comment by jtsmn** [comment]
If lines [705 and 706](https://github.com/pysal/pysal/blob/master/pysal/weights/util.py#L705) read:

```
    if w.id_order.any():
        keys = w.id_order.tolist()
```
I no longer get errors.

**Comment by ljwolf** [comment]
Goodness, I'm really sorry you're hitting lots of rough spots... Definitely trying to get you answers asap. 

I actually don't think a patch is strictly necessary here. According to the [`w_subset` function's docstring](https://github.com/pysal/pysal/blob/master/pysal/weights/Wsets.py#L360):
```
ids                     : list
                          A list containing the IDs to be include in the returned weights
                          object.
```
You've probably used an array. I can replicate using the same data you submitted with #587:
```python
import pysal as ps

weights = ps.queen_from_shapefile('shapefile_for_weights.shp', idVariable='gid')
yellow = ps.pdio.read_files('yellow.shp')

keep_ids = df.gid[~df.total_spen.isnull()].values

w_list = ps.w_subset(weights, keep_ids)
w.full()[0] #succeeds, gives a binary contiguity matrix
w_array = ps.w_subset(weights, keep_ids_array)
w.full()[0] #fails on the error you describe
```
So, immediately, I'd suggest passing a list to `w_subset()`, like recommended in the documentation. Ahh, the vagaries of dynamic typing. Again, sorry this is hitting your use case. 

More generally, we will probably [chat](https://gitter.im/pysal/pysal) about coercing input ids to a list to avoid this.

**Comment by ljwolf** [comment]
Heads up, a fix to this has been filed over in [a feature branch](https://github.com/ljwolf/pysal/tree/pdio).

**Comment by ljwolf** [comment]
I think this is closed by #821 in `pysal/dev`:
```python
import pysal as ps

w = ps.open(ps.examples.get_path('sids2.gal')).read()
w2 = ps.w_subset(w, np.array(w.id_order[0:20]))
w2.full()
```
